### DISCUSSION TITLE
Clang On Linux Shows Much Improved Performance Compared To Clang/LLVM On Windows. Am I Doing Something Wrong?

### DISCUSSION THREAD

**Discussion by 38github** [mikeoliphant/neural-amp-modeler-lv2#51] [General]
I compiled the plugin on Windows using the LLVM package for Visual Studio 2022 and made sure (`cmake.exe -G "Visual Studio 17" -A x64 -T ClangCL ..`) that it actually was Clang that was used during compiling. I am not sure if it used _-O3 -flto=thin_ like I compile it with on Linux but the performance to my surprise was almost like the official VST3 version which is about twice that of my Linux compile.

Anyone know how to successfully add _C_, _CXX_ and _LD_ flags to a cmake project on Windows? I tried _setx _CFLAGS -O3 -flto=thin__ and also adding a couple of lines that I don't remember precisely what they were to the CMakeList file.

Regards

**Comment by 38github** [comment]
I managed to set the CXX flags by inserting `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Ofast -flto=thin")` in `CMakeLists.txt` and a build using around 15 percentage points less CPU was successfully built.

That is not at all comparable to the Linux build though. However, when I added `-march=x86-64-v3` the build is using 45 percentage points less CPU compared to the official NAM VST3 plugin. That is great in my opinion. I also tried x86-64-v2 to see if older CPUs can benefit but it made no real improvement compared to without -march.

**Comment by mikeoliphant** [comment]
When I tested this a while back, as I recall I was seeing about a 10% reduction using clang. So that is roughly in line with what you saw without the ```-march=x86-64-v3```.

I just tested the build you shared on discord, and I'm seeing a 30% improvement over the Visual Studio build (for standard WaveNet models).

My guess is that the official plugin would see comparable improvements. I had a issue open there, but it was closed a while back:

https://github.com/sdatkinson/NeuralAmpModelerPlugin/issues/222

**Reply by 38github** [reply]
I am glad that I can be of help. Try some LSTM models with that build and see an even greater CPU improvement.

**Comment by mikeoliphant** [comment]
@38github The github build should now be producing optimized binaries. I've verified that the Windows version is working. When you get a chance, can you try the Linux version? You can download it here:

https://github.com/mikeoliphant/neural-amp-modeler-lv2/actions/runs/6630279810

**Reply by 38github** [reply]
It is working BUT the CPU is 5.5% while my build is 2.9% using Standard WaveNet.

**Reply by mikeoliphant** [reply]
Yeah - looking at the build, my detection of x64 isn't working on Linux.

What does ```uname -m``` say on your system?

**Reply by mikeoliphant** [reply]
I think I've got it now. Can you try this version:

https://github.com/mikeoliphant/neural-amp-modeler-lv2/actions/runs/6632036018

**Reply by 38github** [reply]
It went down to 3.1% now. Great!

**Comment by ljampietro** [comment]
if cmake outputs this error:

`CMake Error at CMakeLists.txt:3 (project):
  No CMAKE_CXX_COMPILER could be found.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.`

just install clang. and try again.
